src/trace/xray-service.ts

```diff
@@ -0,0 +1,204 @@
+import BigNumber from "bignumber.js";
+import { Socket, createSocket } from "dgram";
+import { SpanContextWrapper } from "./span-context-wrapper";
+import { StepFunctionContext } from "./step-function-service";
+
+const AMZN_TRACE_ID_ENV_VAR = "_X_AMZN_TRACE_ID";
+const AWS_XRAY_DAEMON_ADDRESS_ENV_VAR = "AWS_XRAY_DAEMON_ADDRESS";
+
+interface XrayTraceHeader {
+  traceId: string;
+  parentId: string;
+  sampled: string;
+}
+
+export class XrayService {
+  private header?: string;
+  private context?: XrayTraceHeader;
+  private readonly subsegmentName = "datadog-metadata";
+  private readonly subsegmentNamespace = "datadog";
+  private readonly baggageSubsegmentKey = "root_span_metadata";
+  private readonly subsegmentKey = "trace";
+  private readonly lambdaFunctionTagsKey = "lambda_function_tags";
+
+  constructor() {
+    this.header = process.env[AMZN_TRACE_ID_ENV_VAR];
+    if (this.header === undefined) {
+      logDebug("Couldn't read Xray trace header from env");
+    }
+
+    this.context = this.parseTraceContextHeader();
+    if (this.context === undefined) {
+      logDebug("Couldn't parse Xray trace header from env");
+    }
+  }
+
+  public addLambdaTriggerTags(triggerTags: { [key: string]: string }) {
+    this.add(this.lambdaFunctionTagsKey, triggerTags);
+  }
+
+  public addStepFunctionContext(context: StepFunctionContext) {
+    this.add(this.baggageSubsegmentKey, context);
+  }
+
+  public addMetadata(metadata: Record<string, any>) {
+    this.add(this.subsegmentKey, metadata);
+  }
+
+  private add(key: string, metadata: Record<string, any>) {
+    const subsegment = this.generateSubsegment(key, metadata);
+
+    if (subsegment === undefined) return;
+
+    this.sendSubsegment(subsegment);
+  }
+
+  private generateSubsegment(key: string, metadata: Record<string, any>) {
+    if (this.context === undefined) return;
+
+    const sampled = this.convertToSampleMode(parseInt(this.context.sampled, 10));
+    if (sampled === SampleMode.USER_REJECT || sampled === SampleMode.AUTO_REJECT) {
+      logDebug("Discarding Xray metadata subsegment due to sampling");
+      return;
+    }
+
+    const milliseconds = Date.now() * 0.001;
+
+    return JSON.stringify({
+      id: randomBytes(8).toString("hex"),
+      trace_id: this.context.traceId,
+      parent_id: this.context.parentId,
+      name: this.subsegmentName,
+      start_time: milliseconds,
+      end_time: milliseconds,
+      type: "subsegment",
+      metadata: {
+        [this.subsegmentNamespace]: {
+          [key]: metadata,
+        },
+      },
+    });
+  }
+
+  private parseTraceContextHeader(): XrayTraceHeader | undefined {
+    if (this.header === undefined) return;
+
+    // Example: Root=1-5e272390-8c398be037738dc042009320;Parent=94ae789b969f1cc5;Sampled=1
+    logDebug(`Reading Xray trace context from env var ${this.header}`);
+    const [root, parent, _sampled] = this.header.split(";");
```

Comment:
Careful! The xray trace header can have more than three keys in it!

Reply:
Correct, but in typescript, you can de-structure arrays with fewer constant when an array contains more values to unpack. 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment